COMPILER_ARCHITECTURE.md

```diff
@@ -0,0 +1,1430 @@
+    Fn(Vec<Type>, Box<Type>),
+    
+    /// Type variables (for inference)
+    Var(TypeVar),
+    
+    /// Quantified types (for polymorphism)
+    Forall(Vec<TypeVar>, Box<Type>),
+    
+    /// Units and quantities
+    Quantity {
+        value_type: Box<Type>,  // Integer or Float
+        dimension: Dimension,
+    },
+    
+    /// Trait constraints
+    Constrained {
+        ty: Box<Type>,
+        traits: Vec<TraitRef>,
+    },
+    
+    /// Effect types
+    Effect {
+        result: Box<Type>,
+        effects: Vec<EffectRef>,
+    },
+}
+```
+
+#### Type Inference Algorithm
+
+We implement Algorithm W (Damas-Milner) with constraint generation:
+
+1. **Constraint Generation**: Walk the AST and generate type constraints
+   ```rust
+   fn generate_constraints(expr: &Expr, env: &TypeEnv) -> (Type, Vec<Constraint>)
+   ```
+
+2. **Constraint Solving**: Use unification to solve constraints
+   ```rust
+   fn unify(t1: Type, t2: Type) -> Result<Substitution>
+   ```
+
+3. **Generalization**: Introduce `forall` quantifiers at let-bindings
+   ```rust
+   fn generalize(ty: Type, env: &TypeEnv) -> Type
+   ```
+
+4. **Instantiation**: Replace quantified variables with fresh type variables
+   ```rust
+   fn instantiate(ty: Type) -> Type
+   ```
+
+#### Dimensional Analysis Integration
+
+Unit types are checked during type inference:
+
+```cadenza
+let distance = 100meter
+let time = 5second
+let velocity = distance / time  # Type: Quantity<Float, meter/second>
+
+# Type error: cannot add incompatible dimensions
+let invalid = distance + time  # Error: cannot add meter and second
+```
+
+The type checker maintains dimension equations and solves them alongside type constraints:
+
+```rust
+pub struct DimensionConstraint {
+    left: Dimension,
+    right: Dimension,
+    reason: ConstraintReason,
+}
+```
+
+#### Type Checking Unevaluated Code
+
+When we encounter an unevaluated branch:
+
+1. **Mark as speculative**: Tag the type checking context
+2. **Generate constraints**: Same as evaluated code
+3. **Solve constraints**: Ensure type soundness
+4. **Record result**: Store inferred types but don't generate code
+
+```rust
+impl TypeChecker {
+    fn check_expr(&mut self, expr: &Expr, expected: Type, mode: CheckMode) -> Result<Type> {
+        match mode {
+            CheckMode::Evaluated => {
+                // Normal type checking + code generation
+            }
+            CheckMode::Unevaluated => {
+                // Type checking only, no code generation
+                // Still ensures type soundness
+            }
+        }
+    }
+}
+```
+
+### Type Errors
+
+Type errors include:
+- Expected vs actual type
+- Source location with full context
+- Suggestion for fixes when possible
+- Stack trace showing call chain
+
+---
+
+## Module System
+
+### Module Structure
+
+```cadenza
+# mymodule.cdz
+
+# Private function (not exported by default)
+let internal_helper = fn x -> x + 1
+
+# Public function (explicit export)
+@export
```

Comment:
This is more for exporting artifacts from a running program. So the main use cases for the language I have right now are:
* 3d modeling, kind of like openscad where you define everything in code and then that is able to generate models. So this export would be able to name an object and then be able to be exported as an stl or whatever
* Algorithmic music composition. This export would be for defining a composition that you could listen to
* Making visual art
* Writing interactive books where your program drives simulations or visualizations and describes concepts

Saying this out loud it would be good to put these use cases at the top so it's clear what we're working towards